### PR TITLE
Add more DocTests for `OptionT`

### PR DESCRIPTION
**Pull request by timo-schmid**
#2479

### CODE REVIEW

File: core/src/main/scala/cats/data/OptionT.scala [hunk 1]
```diff
@@ -87,55 +107,212 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
...
   def semiflatMap[B](f: A => F[B])(implicit F: Monad[F]): OptionT[F, B] =
     flatMap(a => OptionT.liftF(f(a)))
 
+  /**
+   * Example:
+   * {{{
+   *  scala> import cats.data.OptionT
+   *  scala> import scala.util.Try
+   *
+   *  scala> // prints "5"
```
Comment (timo-schmid) [inline core/src/main/scala/cats/data/OptionT.scala:180]: Is there a good way to test if the side-effects worked properly?
If not, should we even leave those examples in?

File: core/src/main/scala/cats/data/OptionT.scala [hunk 2]
```diff
@@ -62,6 +62,16 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
    * Catamorphism on the Option. This is identical to [[fold]], but it only has
    * one parameter list, which can result in better type inference in some
    * contexts.
+   *
+   * Example:
+   * {{{
+   * scala> import cats.implicits._
```
Comment (armanbilge) [inline core/src/main/scala/cats/data/OptionT.scala:68]: `cats.implicits` is deprecated.

```suggestion
   * scala> import cats.syntax.all._
```

File: core/src/main/scala/cats/data/OptionT.scala [hunk 3]
```diff
@@ -87,55 +107,212 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
...
       F.imap(value)(_.map(f))(_.map(g))
     }
 
+  /**
+   * Example:
+   * {{{
+   *  scala> import cats.implicits._
+   *  scala> import cats.data.OptionT
+   *
+   *  scala> type ToString[A] = Function1[A, String]
```
Comment (armanbilge) [inline core/src/main/scala/cats/data/OptionT.scala:131]: This is `cats.Show` :)

File: core/src/main/scala/cats/data/OptionT.scala (hunk 1 above)
Comment (armanbilge) [reply core/src/main/scala/cats/data/OptionT.scala:180]: No, and `Try` is not sufficient for suspending side-effects anyway (nothing in Cats is). We should change these examples.

File: core/src/main/scala/cats/data/OptionT.scala [hunk 4]
```diff
@@ -87,55 +107,212 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
...
     semiflatMap(a => F.as(f(a), a))
 
+  /**
+   * Example:
+   * {{{
+   *  scala> import cats.implicits._
+   *  scala> import cats.data.OptionT
+   *
+   *  scala> val optionT: OptionT[List, Int] = OptionT[List, Int](List(Some(2), Some(3), Some(4)))
+   *  scala> optionT.mapFilter(x => Option.when(x % 2 == 0)(x))
```
Comment (armanbilge) [inline core/src/main/scala/cats/data/OptionT.scala:196]: CI might not like `Option.when` on Scala 2.12. I guess we'll find out :)

**Review by armanbilge — COMMENTED** [review]
Thank you so much for this work. Overall looks great, just a few comments. Sorry for the slow review cycle.

I also kicked off CI, I didn't realize it was pending approval.

File: core/src/main/scala/cats/data/OptionT.scala (hunk 1 above)
Comment (satorg) [reply core/src/main/scala/cats/data/OptionT.scala:180]: Actually, `OptionT` does not have to deal with side effects:
```
scala> import cats.data.OptionT, cats.syntax.all._
import cats.data.OptionT
import cats.syntax.all._

scala> val optt = OptionT.some[Either[String, *]](3)
val optt: cats.data.OptionT[[β$0$]scala.util.Either[String,β$0$],Int] = OptionT(Right(Some(3)))

scala> optt.semiflatTap { case 1 | 2 | 3 => Right("hit!"); case _ => Left("miss!") }
val res0: cats.data.OptionT[[β$0$]scala.util.Either[String,β$0$],Int] = OptionT(Right(Some(3)))

scala> optt.semiflatTap { case 0 | 1 | 2 => Right("hit!"); case _ => Left("miss!") }
val res1: cats.data.OptionT[[β$0$]scala.util.Either[String,β$0$],Int] = OptionT(Left(miss!))

```

File: core/src/main/scala/cats/data/OptionT.scala [hunk 5]
```diff
@@ -87,55 +107,212 @@ final case class OptionT[F[_], A](value: F[Option[A]]) {
...
    * with an effect returning `None` as argument.
+   * 
+   * Example:
+   * {{{
+   *  scala> import cats.data.OptionT
+   *  scala> import scala.util.Try
+   *
+   *  scala> // prints "no value"
+   *  scala> val optionT: OptionT[Try, Int] = OptionT[Try, Int](Try(None))
+   *  scala> optionT.flatTapNone(Try(println("no value"))).value
```
Comment (satorg) [inline core/src/main/scala/cats/data/OptionT.scala:284]: I think this example is also not really good for the same reason as one for `semiflatTap`.

File: core/src/main/scala/cats/data/OptionT.scala (hunk 1 above)
Comment (timo-schmid) [reply core/src/main/scala/cats/data/OptionT.scala]: That's very clever, I added it to the PR! :+1:

File: core/src/main/scala/cats/data/OptionT.scala (hunk 2 above)
Comment (timo-schmid) [reply core/src/main/scala/cats/data/OptionT.scala]: Thanks, that was news to me. There were actually a bunch of unused `cats.implicits._` imports in the Doctests.
I removed them.

File: core/src/main/scala/cats/data/OptionT.scala (hunk 3 above)
Comment (timo-schmid) [reply core/src/main/scala/cats/data/OptionT.scala]: Of course :facepalm: - I was so happy I finally found an example for this I didn't even notice.
Updated.

File: core/src/main/scala/cats/data/OptionT.scala (hunk 4 above)
Comment (timo-schmid) [reply core/src/main/scala/cats/data/OptionT.scala]: Correct, I changed to `Option(...).filter(...)`.
I assume with `~+coreJVM/testOnly **OptionTDoctest` (added the `+`) I should catch those problems, right?

Comment (armanbilge) [reply core/src/main/scala/cats/data/OptionT.scala]: Possibly. I don't know how `~` and `+` interact in sbt.

**Pull request merged**